src/main/java/org/jenkinsci/plugins/workflow/support/actions/LogActionImpl.java

```diff
@@ -149,12 +153,10 @@ public FlowNode getParent() {
      */
     private static boolean isRunning(FlowNode node) {
         if (node instanceof BlockStartNode) {
-            for (FlowNode head : node.getExecution().getCurrentHeads()) {
-                if (new LinearBlockHoppingScanner().findFirstMatch(head, Predicates.equalTo(node)) != null) {
-                    return true;
-                }
-            }
-            return false;
+            // Block start is considered running if currently executing nodes are part of the block
+            List<FlowNode> headNodes = node.getExecution().getCurrentHeads();
+            AbstractFlowScanner scanner = (headNodes.size() > 1) ? new DepthFirstScanner() : new LinearScanner();
```

Comment:
Yes, it is still O(n)ish and not amazing, but still much better with many parallels.  The "right" way is to store a datastructure holding the block structure of the flow that can be efficiently queried for enclosing blocks.

We've talked about that before, but let's not hold up a solid fix just because we want to reach an even better solution that may be much further off.